storage/allocator.go

```diff
@@ -35,21 +37,16 @@ const (
 	// target and it will always be eligible to rebalance replicas to other
 	// stores.
 	maxFractionUsedThreshold = 0.95
-	// minFractionUsedThreshold: if the mean fraction used of a list of store
-	// descriptors is less than this, then range count will be used to make
-	// rebalancing decisions instead of the fraction of bytes used. This is
-	// useful for distributing load evenly on nascent deployments.
-	minFractionUsedThreshold = 0.02
-	// rebalanceFromMean is used to declare a range above and below the average
-	// used capacity of the cluster. If a store's usage is below this range, it
-	// is a rebalancing target and can accept new replicas; if usage is above
-	// this range, the store is eligible to rebalance replicas to other stores.
-	rebalanceFromMean = 0.025 // 2.5%
 
 	// priorities for various repair operations.
 	removeDeadReplicaPriority  float64 = 10000
 	addMissingReplicaPriority  float64 = 1000
 	removeExtraReplicaPriority float64 = 100
+
+	// minSecsBetweenRebalances is the minimum number of seconds that must elapse
+	// between replica rebalances. This should be slightly larger than
+	// server.gossipStoresInterval.
+	minRebalanceInterval = 65 * time.Second
```

Comment:
I talked to Pete in person, but it's useful to document some of that here (and the rest in an RFC).

There's a tension here between avoiding flip-flopping of replicas between nodes and speeding things up. Because we only gossip `StoreDescriptor`s once a minute, there's always a risk that the allocator is using outdated data, which causes something like a feedback loop.

Let's take your example:

1. Stores 1, 2, 3, each have 100 replicas
2. You add store 4, which is empty.
3. Within the first ~1 minute after adding store 4, the allocators for stores 1-3 will think that store 4 is empty based on the gossiped `StoreDescriptor`.
4. Store 4 gets too many replicas (again, because it appears to have 0 replicas based on outdated gossip info)
5. Store 4 needs to rebalance replicas back to stores 1-3 to even things out.
6. Lots of flip-flopping of replicas happens.
7. In my own experiments, it takes several *thousand* range movements for things to stabilize in this scenario. Using the slower but surer heuristic implemented here, things stabilize in < 100 range events.